### PR TITLE
per crate min version

### PR DESCRIPTION
**Pull request by bilelmoussaoui**
Fixes https://github.com/gtk-rs/gir/issues/1231
Fixes https://github.com/gtk-rs/gir/issues/1099

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -8,7 +8,7 @@ authors = [
 ]
 build = "build.rs"
 exclude = ["Gir*.toml", "tests/**/*", "*.md", "generator.py"]
-edition = "2021"
+edition = "2018"
```
Comment (GuillaumeGomez) [inline Cargo.toml:11]: Why reverting this?

Comment (GuillaumeGomez) [inline Cargo.toml:11]: Why reverting this?

File: src/config/external_libraries.rs [hunk 2]
```diff
@@ -25,17 +29,35 @@ pub fn read_external_libraries(toml: &toml::Value) -> Result<Vec<ExternalLibrary
...
         for custom_lib in custom_libs {
-            let crate_name = custom_lib.0;
-            if let Some(namespace) = custom_lib.1.as_str() {
+            if let Some(info) = custom_lib.1.as_table() {
+                println!("{:#?}", info);
+                let namespace = custom_lib.0.as_str();
+                let crate_name_ = info
+                    .get("crate")
+                    .map(|c| c.as_str().expect("crate name must be a string").to_string())
+                    .unwrap_or_else(|| crate_name(&namespace));
```
Comment (GuillaumeGomez) [inline src/config/external_libraries.rs:38]:
```suggestion
                    .unwrap_or_else(|| crate_name(namespace));
```

**Comment by sophie-h** [comment]
Can't reproduce that adding 

```
[external_libraries]
Gio = {min_version = "2.32"}
```

fixes libadwaita

File: Cargo.toml (hunk 1 above)
Comment (bilelmoussaoui) [reply Cargo.toml:11]: For rust analyzer, I wasn't meaning to push it

Comment (sdroege) [reply Cargo.toml:11]: Why? For me it works with 2021 :)

Comment (GuillaumeGomez) [reply Cargo.toml:11]: Question remains. :)

Comment (GuillaumeGomez) [reply Cargo.toml:11]: Still waiting.

**Comment by bilelmoussaoui** [comment]
I will re-generate the various crates tomorrow and add some docs to the book to explain the new options.

**Comment by sdroege** [comment]
Why does this need changes in the existing crates?

**Comment by bilelmoussaoui** [comment]
> Why does this need changes in the existing crates?

Because gtk3/gtk4 needs to set the new gio features or define a min version and regen

**Review by sdroege — APPROVED** [review]
Looks good to me

**Comment by sdroege** [comment]
@bilelmoussaoui Did you check if this works correctly with gstreamer-rs? In theory nothing should be needed there except for maybe adding a minimum GLib/GIO version.

**Comment by bilelmoussaoui** [comment]
> @bilelmoussaoui Did you check if this works correctly with gstreamer-rs? In theory nothing should be needed there except for maybe adding a minimum GLib/GIO version.

Yes, I will be sending a regen PR in a minute. Doing the libadwaita/libhandy ones

**Comment by sdroege** [comment]
@bilelmoussaoui Ok then let's get this one here in so you can do the other PRs properly :)

**Pull request merged**